File: Langite/Syntax/Lexer.cs
using System;
using System.Diagnostics;

namespace Langite.Syntax
{
    public sealed class Lexer
    {
        private int _column = 1;
        private int _line = 1;

        private int _position = 0;

        public Lexer(string filepath, string source)
        {
            Filepath = filepath;
            Source = source;
        }

        public string Filepath { get; }
        public string Source { get; }

        private char CurrentChar => _position < Source.Length
            ? Source[_position]
            : '\0';

        public Token PeekToken()
        {
            var position = _position;
            var line = _line;
            var column = _column;
            var token = NextToken();
            _position = position;
            _line = line;
            _column = column;
            return token;
        }

        public TokenKind PeekKind()
        {
            return PeekToken().Kind;
        }

        public Token NextToken()
        {
            Start:
            while (char.IsWhiteSpace(CurrentChar) && CurrentChar != '\n')
                NextChar();

            var startLocation = new SourceLocation(Filepath, _position, _line, _column);

            if (CurrentChar == '"')
            {
                NextChar();

                var value = string.Empty;
                while (CurrentChar != '"' && CurrentChar != '\0')
                    if (CurrentChar == '\\')
                    {
                        NextChar();
                        value += CurrentChar switch
                        {
                            '0' => '\0',
                            't' => '\t',
                            'n' => '\n',
                            'r' => '\r',
                            _ => throw new CompileError(new SourceLocation(Filepath, _position, _line, _column),
                                $"Unknown escape character {NextChar()}"),
                        };
                    }
                    else
                    {
                        value += NextChar();
                    }

                var closeQuote = NextChar();
                if (closeQuote != '"')
                {
                    Debug.Assert(closeQuote == '\0');
                    throw new CompileError(startLocation, "String literal unclosed at end of file");
                }

                return new Token(TokenKind.String, startLocation, _position - startLocation.Position, value);
            }

            if (char.IsAscii(CurrentChar) && char.IsDigit(CurrentChar))
            {
                var @base = 10;
                if (CurrentChar == '0')
                {
                    NextChar();
                    switch (CurrentChar)
                    {
                        case 'b':
                            NextChar();
                            @base = 2;
                            break;

                        case 'o':
                            NextChar();
                            @base = 8;
                            break;

                        case 'd':
                            NextChar();
                            @base = 10;
                            break;

                        case 'x':
                            NextChar();
                            @base = 16;
                            break;
                    }
                }

                var intValue = 0L;

                while (char.IsAscii(CurrentChar) && char.IsLetterOrDigit(CurrentChar) || CurrentChar == '_')
                {
                    if (CurrentChar == '_')
                    {
                        NextChar();
                        continue;
                    }

                    var chr = CurrentChar;
                    var value = chr switch
                    {
                        >= '0' and <= '9' => (long) chr - '0',
                        >= 'A' and <= 'Z' => (long) chr - 'A' + 10,
                        >= 'a' and <= 'z' => (long) chr - 'a' + 10,
                        _ => throw new ArgumentOutOfRangeException(),
                    };

                    if (value >= @base)
                    {
                        var location = new SourceLocation(Filepath, _position, _line, _column);
                        NextChar();
                        throw new CompileError(location, $"Digit '{chr}' is too big for base {@base}");
                    }

                    intValue *= @base;
                    intValue += value;

                    NextChar();
                }

                if (CurrentChar == '.')
                {
                    var floatValue = (double) intValue;
                    var discriminant = 1L;

                    while (char.IsAscii(CurrentChar) && char.IsLetterOrDigit(CurrentChar) || CurrentChar == '_')
                    {
                        if (CurrentChar == '_')
                        {
                            NextChar();
                            continue;
                        }

                        var chr = CurrentChar;
                        var value = chr switch
                        {
                            >= '0' and <= '9' => (long) chr - '0',
                            >= 'A' and <= 'Z' => (long) chr - 'A' + 10,
                            >= 'a' and <= 'z' => (long) chr - 'a' + 10,
                            _ => throw new ArgumentOutOfRangeException(),
                        };

                        if (value >= @base)
                        {
                            var location = new SourceLocation(Filepath, _position, _line, _column);
                            NextChar();
                            throw new CompileError(location, $"Digit '{chr}' is too big for base {@base}");
                        }

                        discriminant *= @base;
                        floatValue += (double) value / discriminant;

                        NextChar();
                    }

                    return new Token(TokenKind.Float, startLocation, _position - startLocation.Position, floatValue);
                }

                return new Token(TokenKind.Integer, startLocation, _position - startLocation.Position, intValue);
            }

            if (char.IsLetterOrDigit(CurrentChar) || CurrentChar == '_')
            {
                while (char.IsLetterOrDigit(CurrentChar) || CurrentChar == '_')
                    NextChar();
                var length = _position - startLocation.Position;
                var name = Source.Substring(startLocation.Position, length);
                return new Token(TokenKindMethods.FromString(name), startLocation, length, name);
            }

            {
                var chr = NextChar();

                if (chr == '/' && CurrentChar == '/')
                {
                    while (CurrentChar != '\n')
                        NextChar();
                    goto Start;
                }

                if (chr == '/' && CurrentChar == '*')
                    throw new NotImplementedException();
                // goto Start;

                {
                    if (TokenKindMethods.FromDoubleChar(chr, CurrentChar) is { } kind)
                    {
                        NextChar();
                        return new Token(kind, startLocation, _position - startLocation.Position);
                    }
                }
                {
                    if (TokenKindMethods.FromChar(chr) is { } kind)
                        return new Token(kind, startLocation, _position - startLocation.Position);
                }

                throw new CompileError(startLocation, $"Unexpected character '{chr}'");
            }
        }

        private char NextChar()
        {
            var current = CurrentChar;
            if (current != '\0')
            {
                _position++;
                _column++;
                if (current == '\n')
                {
                    _line++;
                    _column = 1;
                }
            }

            return current;
        }
    }
}

File: src/syntax/lexer.rs
use std::str::Chars;

use derive_more::Display;

use super::{SourceLocation, SourceSpan, Token, TokenKind};

#[derive(Debug, Display)]
pub enum LexerError<'filepath> {
    #[display(fmt = "{location}: Unexpected character {character:?}")]
    UnexpectedCharacter {
        location: SourceSpan<'filepath>,
        character: char,
    },
    #[display(fmt = "{location}: Unclosed block comment")]
    UnclosedBlockComment {
        location: SourceSpan<'filepath>,
    },
    InvalidIntegerLiteral {
        location: SourceSpan<'filepath>,
    },
}

#[derive(Clone)]
pub struct Lexer<'filepath, 'source> {
    filepath: &'filepath str,
    source: &'source str,
    chars: Chars<'source>,
    location: SourceLocation,
}

impl<'filepath, 'source> Lexer<'filepath, 'source> {
    pub fn new(filepath: &'filepath str, source: &'source str) -> Self {
        Lexer {
            filepath,
            source,
            chars: source.chars(),
            location: SourceLocation {
                position: 0,
                line: 1,
                column: 1,
            },
        }
    }

    pub fn get_filepath(&self) -> &'filepath str {
        self.filepath
    }

    pub fn next_char(&mut self) -> Option<char> {
        let current = self.chars.next()?;
        self.location.position += current.len_utf8();
        self.location.column += 1;
        if current == '\n' {
            self.location.line += 1;
            self.location.column = 1;
        }
        Some(current)
    }

    pub fn peek_char(&mut self) -> Option<char> {
        self.clone().next_char()
    }

    pub fn next_token(
        &mut self,
    ) -> Result<Option<Token<'filepath, 'source>>, LexerError<'filepath>> {
        loop {
            let start_location = self.location;
            return Ok(Some(Token {
                kind: match self.next_char() {
                    None => return Ok(None),
                    Some('\n') => TokenKind::Newline,
                    Some('/') if self.peek_char() == Some('/') => {
                        while self.peek_char() != Some('\n') {
                            self.next_char();
                        }
                        continue;
                    }
                    Some('/') if self.peek_char() == Some('*') => {
                        self.next_char();
                        let mut depth = 1usize;
                        while let Some(c) = self.next_char() {
                            if c == '/' && self.peek_char() == Some('*') {
                                self.next_char();
                                depth += 1;
                            } else if c == '*' && self.peek_char() == Some('/') {
                                self.next_char();
                                depth -= 1;
                            }
                        }
                        if depth > 0 {
                            return Err(LexerError::UnclosedBlockComment {
                                location: SourceSpan {
                                    filepath: self.filepath,
                                    start: start_location,
                                    end: self.location,
                                },
                            });
                        }
                        continue;
                    }
                    Some('(') => TokenKind::OpenParenthesis,
                    Some(')') => TokenKind::CloseParenthesis,
                    Some('{') => TokenKind::OpenBrace,
                    Some('}') => TokenKind::CloseBrace,
                    Some(',') => TokenKind::Comma,
                    Some(':') => TokenKind::Colon,
                    Some('=') if self.peek_char() == Some('>') => {
                        self.next_char();
                        TokenKind::FatRightArrow
                    }
                    Some('+') => TokenKind::Plus,
                    Some('-') => TokenKind::Minus,
                    Some('*') => TokenKind::Asterisk,
                    Some('/') => TokenKind::Slash,
                    Some('"') => {
                        todo!()
                    }
                    Some(c) if c.is_ascii_digit() => {
                        while let Some(c) = self.peek_char() {
                            if !c.is_ascii_digit() {
                                break;
                            }
                            self.next_char();
                        }
                        let integer = &self.source[start_location.position..self.location.position];
                        TokenKind::Integer(integer.parse().map_err(|_| {
                            LexerError::InvalidIntegerLiteral {
                                location: SourceSpan {
                                    filepath: self.filepath,
                                    start: start_location,
                                    end: self.location,
                                },
                            }
                        })?)
                    }
                    Some(c) if c.is_alphabetic() || c == '_' => {
                        while let Some(c) = self.peek_char() {
                            if !c.is_ascii_alphanumeric() && c != '_' {
                                break;
                            }
                            self.next_char();
                        }
                        let name = &self.source[start_location.position..self.location.position];
                        match name {
                            "proc" => TokenKind::ProcKeyword,
                            "return" => TokenKind::ReturnKeyword,
                            "do" => TokenKind::DoKeyword,
                            "let" => TokenKind::LetKeyword,
                            "var" => TokenKind::VarKeyword,
                            _ => TokenKind::Name(name),
                        }
                    }
                    Some(c) if c.is_whitespace() => continue,
                    Some(c) => {
                        return Err(LexerError::UnexpectedCharacter {
                            location: SourceSpan {
                                filepath: self.filepath,
                                start: start_location,
                                end: self.location,
                            },
                            character: c,
                        });
                    }
                },
                span: SourceSpan {
                    filepath: self.filepath,
                    start: start_location,
                    end: self.location,
                },
            }));
        }
    }

    pub fn peek_token(&self) -> Result<Option<Token<'filepath, 'source>>, LexerError<'filepath>> {
        self.clone().next_token()
    }
}

impl<'filepath, 'source> Iterator for Lexer<'filepath, 'source> {
    type Item = Result<Token<'filepath, 'source>, LexerError<'filepath>>;

    fn next(&mut self) -> Option<Self::Item> {
        self.next_token().transpose()
    }
}
